### ISSUE TITLE
Wiki Page Update for Adding Authentication

### ISSUE THREAD

**Issue by csexton** [Shopify/dashing#370]
I wanted to propose a change to the wiki page, but get some feedback. Seems like an issue is the best way to go about that.

The instructions for Google Apps Authentication show a [deprecated method](https://developers.google.com/accounts/docs/OAuth). Since that no longer works for new apps, I tinkered around with OAuth 2 for a bit and came up with the following. Should we up date the contents on [How-to: Add authentication](https://github.com/Shopify/dashing/wiki/How-to:-Add-authentication)?

``` ruby
require 'openid/store/filesystem'
require 'omniauth/strategies/google_apps'
require 'omniauth-google-oauth2'
require 'dashing'

configure do
  set :auth_token, 'YOUR_TOKEN'

  helpers do
    def protected!
      if ENV["GOOGLE_CLIENT_ID"]
        redirect '/auth/g' unless session[:user_id]
      end
    end
  end

  use Rack::Session::Cookie
  use OmniAuth::Builder do
    provider OmniAuth::Strategies::GoogleOauth2, ENV["GOOGLE_CLIENT_ID"], ENV["GOOGLE_CLIENT_SECRET"],
      :store => OpenID::Store::Filesystem.new('./tmp'),
      :name => 'g',
      :domain => 'radiusnetworks.com',
      :provider_ignores_state => true
  end

  get '/auth/g/callback' do
    if auth = request.env['omniauth.auth']
      if request.env["omniauth.auth"]['extra']['raw_info']['hd'] == "YOURDOMAIN.com"
        session[:user_id] = auth['info']['email']
        redirect '/'
      else
        redirect '/auth/bad'
      end
    else
      redirect '/auth/failure'
    end
  end

  get '/auth/failure' do
    'Nope.'
  end

  get '/auth/bad' do
    'ZOMG! You must use the YOURDOMAIN.'
  end
end

map Sinatra::Application.assets_prefix do
  run Sinatra::Application.sprockets
end

run Sinatra::Application
```

If you'd rather I can just write up a blog post with the details.

**Comment by olivierlefloch** [comment]
+1

**Comment by joshpuetz** [comment]
I just successfully set this up, and here are my general notes:

Here are the gems I had to add to support csexton's excellent post (https://github.com/Shopify/dashing/issues/370#issue-34923544):

```
gem install "omniauth-openid"
gem install "omniauth-google-apps"
gem install "omniauth-google-oauth2"
```

Then, you'll need to create a project on Google's Developer Console, give it access to the Google+ API, and create an oAuth service account for your web application (that's what will generate a `CLIENT_ID` and `CLIENT_SECRET` to your in your application). In addition, you need to set the callback location of this client to be `<your domain>/auth/g/callback`

Instructions on this process are _kinda_ spelled out here: https://developers.google.com/console/help/new/?hl=en_US#serviceaccounts

**Comment by abepetrillo** [comment]
I've just gone through a similar process to get it working. Here's what I ended up with

```ruby
require 'omniauth/strategies/google_oauth2'
require 'dashing'
require 'yaml'

configure do
  set :access_token, ENV['AUTH_TOKEN']
  if Sinatra::Base.production?
    helpers do

      def protected!
        redirect '/auth/google_oauth2' unless session[:user_id]
      end

    end

    use Rack::Session::Cookie, secret: ENV['secret']


    use OmniAuth::Builder do
      provider :google_oauth2, ENV['GOOGLE_CLIENT_ID'], ENV["GOOGLE_CLIENT_SECRET"]
    end

    get '/auth/google_oauth2/callback' do
      if auth = request.env['omniauth.auth']
        session[:user_id] = auth['info']['email']
        redirect '/abe'
      else
        redirect '/auth/failure'
      end
    end

    get '/auth/failure' do
      'Nope.'
    end
  end
end
map Sinatra::Application.assets_prefix do
  run Sinatra::Application.sprockets
end

run Sinatra::Application

```

**Comment by snario** [comment]
OpenID 2.0 for Google accounts is going away.
Developers should migrate to OpenID Connect by April 20, 2015.

**Comment by csexton** [comment]
@snario according to Google's documentation on [Migrating from OpenID 2.0 to OpenID Connect](https://developers.google.com/identity/protocols/OpenID2Migration) it seems that "OpenID Connect" **is** "OAuth 2.0".

Their [OpenID Connect](https://developers.google.com/identity/protocols/OpenIDConnect) docs say:

> Google's OAuth 2.0 APIs can be used for both authentication and authorization. This document describes our OAuth 2.0 implementation for authentication, which conforms to the OpenID Connect specification.

Srsly, Who picks these names? This is confusing as hell.

I believe everything we are suggesting here has to do with OAuth 2.0 (and hopefully would work with OpenID Connect), but am not certain how to find out if that is the case.

**Comment by snario** [comment]
Hopefully? I just set up my dashboard with the code in the latest comment
above and when I navigated to my login page I had that message pop up.
ᐧ


On Thu, Apr 9, 2015 at 2:21 PM, Christopher Sexton <notifications@github.com

**Comment by EiNSTeiN-** [comment]
Please don't use (or give examples of code using) `:provider_ignores_state => true`. This will make your application vulnerable to [CSRF](http://www.twobotechnologies.com/blog/2014/02/importance-of-state-in-oauth2.html) during the oauth process.

**Comment by tylermauthe** [comment]
I agree. It seems that using Omniauth State requires you to implement a CSRF mechanism:
https://github.com/zquestz/omniauth-google-oauth2#omniauth-state
